### PR TITLE
idmap: support skipping directories

### LINKED ISSUE #4690: volume attach fails when ZFS snapshots exists and snapdir=visible
# Required information

 * Distribution: Ubuntu
 * Distribution version: 18.04
 * The output of "lxc info" or if that fails:
   * Kernel version: 4.15.0-23-generic
   * LXC version: 3.0.0
   * LXD version: 3.0.0
   * Storage backend in use: ZFS

# Issue description

I'm trying to attach a  custom volume to a container and it fails with: "Error: Failed to setup device: 

> Failed to initialize storage volume "test-shared-data" of type "custom" on storage pool "default": Failed to change ownership of: /var/lib/lxd/storage-pools/default/custom/test-shared-data/.zfs/snapshot/cvlt."

This happens when I set the ZFS snapdir=visible attribute  which I had to do the allow CommVault backup agent to drill through the snapshot directory to backup a coherent state ( the 'cvlt' snapshot)

# Steps to reproduce

 lxc storage volume create default test-shared-data
 zfs set snapdir=visible zfspool/lxd/custom/test-shared-data
 zfs snapshot zfspool/lxd/custom/test-shared-data@cvlt
 lxc storage volume attach default test-shared-data testcontainer /shared-data

### PR DESCRIPTION
**Pull request by brauner**
Closes #4690.

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

### CODE REVIEW

**Comment by stgraber** [comment]
Looks good to me but I wonder if a skip function would be more flexible in the future, effectively passing a function to the shift call which if set and returns true for a given path will have the shift skipped.

What do you think?

**Comment by brauner** [comment]
@stgraber, would make sense too and is not a difficult change.

File: shared/idmap/idmapset_linux.go [hunk 1]
```diff
@@ -483,6 +484,18 @@ func (set *IdmapSet) doUidshiftIntoContainer(dir string, testmode bool, how stri
...
 		}
 
+		if skipDirs != nil && len(skipDirs) > 0 {
+			strippedPath := path
+			if dir != "" {
+				strippedPath = path[len(dir):]
+			}
+
+			if fi.IsDir() && shared.StringInSlice(strippedPath, skipDirs) {
+				logger.Errorf("SKIPPING: %s", strippedPath)
```
Comment (brauner) [inline shared/idmap/idmapset_linux.go]: Ups, that needs to be removed.

**Comment by brauner** [comment]
@stgraber

**Pull request merged**